lib/srv/db/sqlserver/kinit/kinit.go

```diff
@@ -0,0 +1,210 @@
+	return &CommandLineKInit{
+		authClient:      authClient,
+		userName:        user,
+		cacheName:       fmt.Sprintf("%s@%s", user, realm),
+		realmName:       realm,
+		kdcHostName:     kdcHost,
+		adminServerName: adminServer,
+
+		ldapCertificate: ldapCA,
+
+		certPath: fmt.Sprintf("%s.pem", user),
+		keyPath:  fmt.Sprintf("%s-key.pem", user),
+		log:      logrus.StandardLogger(),
+	}
+}
+
+// CreateOrAppendCredentialsCache creates or appends to an existing credentials cache.
+func (k *CommandLineKInit) CreateOrAppendCredentialsCache(ctx context.Context) error {
+
+	tmp := filepath.Join(os.TempDir(), "kinit")
+	err := os.MkdirAll(tmp, 0664)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	certPath := filepath.Join(tmp, fmt.Sprintf("%s.pem", k.userName))
+	keyPath := filepath.Join(tmp, fmt.Sprintf("%s-key.pem", k.userName))
+	userCAPath := filepath.Join(tmp, "userca.pem")
+
+	clusterName, err := k.authClient.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	certPEM, keyPEM, err := windows.CertKeyPEM(ctx, k.userName, k.realmName, time.Second*60*60, clusterName.GetClusterName(), windows.LDAPConfig{
+		Addr:               k.kdcHostName,
+		Domain:             k.realmName,
+		Username:           k.userName,
+		InsecureSkipVerify: false,
+		ServerName:         k.adminServerName,
+		CA:                 k.ldapCertificate,
+	}, k.authClient)
+
+	userCA, err := k.authClient.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       types.UserCA,
+		DomainName: clusterName.GetClusterName(),
+	}, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// get the user CA certificate bytes
+	var caCert []byte
+	keyPairs := userCA.GetActiveKeys().TLS
+	for _, keyPair := range keyPairs {
+		if keyPair.KeyType == types.PrivateKeyType_RAW {
+			caCert = keyPair.Cert
+		}
+	}
+
+	if caCert == nil {
+		return trace.Wrap(errors.New("no certificate authority was found in userCA active keys"))
+	}
+
+	// store files in temp dir
+	err = os.WriteFile(certPath, certPEM, 0644)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = os.WriteFile(keyPath, keyPEM, 0644)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = os.WriteFile(userCAPath, caCert, 0644)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	krbConfPath := filepath.Join(tmp, fmt.Sprintf("krb_%s", k.userName))
+	err = k.WriteKRB5Config(krbConfPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// I'm not a fan of this or of writing all these files in the first place, but in testing kinit, it does not have
+	// the ability to access key data over stdin and requires these files to function for x509 auth
+	defer func() {
+		_ = os.RemoveAll(tmp)
+	}()
+
+	cmd := exec.CommandContext(ctx,
+		"kinit",
+		"-X", fmt.Sprintf("X509_anchors=FILE:%s", userCAPath),
+		"-X", fmt.Sprintf("X509_user_identity=FILE:%s,%s", certPath, keyPath), k.userName,
+		"-c", k.cacheName)
```

Comment:
cacheName is a path to the cache where current dir will be used as default. We should make sure that we are creating cache in the known directory.